### ISSUE TITLE
[Monsterhearts-2] Translations.

### ISSUE THREAD

**Issue by WallaceMcGregor** [Roll20/roll20-character-sheets#4822]
Hi.

Could be possible to upgrade the **Monsterhearts-2** sheet by @Lockbox313 to allow translations?

The spanish version will be released in a few months and I would like to translate the sheet.

Thanks.

**Comment by Anduh** [comment]
I had my [ACSI script](https://github.com/Anduh/acsi) autogenerate i18n-tags for the sheet, so if you can find a Pro member to check how good of a result it gave and generate the translation.json file, the person could have most work done for him/her.

[Autogenerated result: Monsterhearts-2.html](https://gist.github.com/Anduh/02977074f79f76dd2989d88fa3065863)

**Comment by WallaceMcGregor** [comment]
Thanks.

I don't know if the sheet contributors (@Lockbox313 and @TripsLetash ) are still "online". I am just a Plus member and I do not know too much about programming.

**Comment by TripsLetash** [comment]
I'm online but I haven't touched these sheets in over a year. You're welcome to make the necesarry edits as you see fit. I don't even think I'm still a plus member on roll20

**Comment by Lockbox313** [comment]
Here is the result https://app.roll20.net/join/4072362/j60XaA of Anduhs auto generator.  The generator stripped the  > from </button/>.  It is also doing some wierd things with <b> tags seems to be inserting  the data-i18n in the opening and closing tag.

Also any chance you could output a default translation.json from my dim memories of the DOM it should just be a matter of grabbing the innerHTML of the tag you are inserting the data-i18n into .

**Comment by Lockbox313** [comment]
The generator did not seem to like the "type="number" in the button either.

**Comment by Anduh** [comment]
Yeah the script is crude so its a bit hit and miss if the output is worth using. I'll see if I can fix these newfound errors. Having it create a `translation.json` file is one of the features considered after the core functionality is more reliable.

**Comment by WallaceMcGregor** [comment]
Also, I've seen that when you change the playbook the sheet don't load . It shows only  the default playbook (The Fae).

That's on the game @Lockbox313 linked.

**Comment by Lockbox313** [comment]
The intent of that was to show the results of the auto generator, not to hand you a full functioning sheet with translation tags at the drop of a hat.

This is still a lot of work. There is a lot of malformed html in the original sheet with tags out of place and random </p> tags inserted.  The generator removes the ending from some or all of the </button>, </h1> and </b>.  Understandably the generator also has problems with large blocks of text with <i> and <b> and other text formatting tags. It also has an issue with drop down options which I have not figured out.

I have just spent a couple of hours going through and fixing the structure and mangled tags and replacing the text of the moves with the original.  The text of the moves may be an issue as I do not know how Roll20 translator works with large blocks of text.

A  basic translation.json needs created as well which means going through and pulling all the data-i18n and the associated text.

This is all the time I have for this currently this. Here is the sheet as it stands now. [https://github.com/Lockbox313/Monsterhearts/blob/master/Monsterhearts-2.html](url)

**Comment by Anduh** [comment]
So it sounds like this time it was barely a timesaver in comparison to just doing it from scratch?
I'm sorry the output wasn't really worth it this time, but I warned that:
>If the script didn't create a timesaving starting point, it doesn't have to be used.

**Comment by Lockbox313** [comment]
The biggest issue is the mangling of tags

**Comment by Anduh** [comment]
Yeah ofc, but was it still better than doing it manually from scratch?

**Comment by Lockbox313** [comment]
Only if you fix what is breaking the tags.

**Comment by Anduh** [comment]
After improving the ACSI script(breaks less tags now) I've now updated the [html and translation.json](https://gist.github.com/Anduh/02977074f79f76dd2989d88fa3065863) it creates, and while it's a noticeable improvement, it still have hickups that need to be manually. Might in a few weeks do the last cleanup my generated translation requires for us to get a working translation.

**Comment by WallaceMcGregor** [comment]
Thanks. Waiting for your "cleanup" to begin the translation.

**Comment by JBeighel** [comment]
I used a lot of the Monsterhearts-2 sheet HTML for the Tremulus one.  Assuming the translation work I did there works, I could use the same techniques to get i18n tags in the Monsterhearts sheet.

@WallaceMcGregor do you think you could let me know if that sheet is reasonably translatable?  I don't want to start that effort without knowing if the results will be useful.

**Comment by WallaceMcGregor** [comment]
> @WallaceMcGregor do you think you could let me know if that sheet is reasonably translatable? I don't want to start that effort without knowing if the results will be useful.

@JBeighel the modification to allow translations is not ended or committed (several months have passed without news). I don't know how is or if It'll work.

**Comment by Lockbox313** [comment]
@WallaceMcGregor  @Anduh  posted revised html and  translation,josn onMay 15.  If you were keen to see this resolved  you could have taken these files and the CSS from the the Roll20 repository and setup a game with a custom sheet and tested it  for yourself and the committed it yourself.

**Comment by WallaceMcGregor** [comment]
@Lockbox313 I am not a Pro user so I can't use a custom sheet or upload one.

**Comment by Lockbox313** [comment]
@WallaceMcGregor  You won't spend $10 on a PRO membership for a month to help with this but you expect others who have to work for you for free and then complain when they are not fast enough for you.

**Comment by WallaceMcGregor** [comment]
@Lockbox313 I've never complained anyone about the time to doing something. I've never said something like that. I am not a programer and I've made a lot of sheets translations to my language (including the Roll20 interface) or reporting errors, only trying to contribute the best I can. I am not a english native speaker so I think maybe could be a misinterpretation if sometime sounded like that.

**Comment by Lockbox313** [comment]
@WallaceMcGregor "several months have passed without news"

**Comment by WallaceMcGregor** [comment]
I'm just answering @JBeighel If you think "several months have passed without news" is a complaint I can say nothing. I am trying to explain him the state. Anyway If you think that simple comment is a complaint, I don't know what to say you. That wasn't my intention. I've never thought it could be taken as a complain. I don't get it.

**Comment by Anduh** [comment]
The translated code I posted was autogenerated by a script, and wasn't fairing well for this particular sheet  so it's understandable that it's unlearn if it's better to start from the actual sheet or from the result of my script. I have little to no time to participate unfortunately, but can run my script again and see if the result is better since I might have improved it since last time.

But yeah I was a bit unclear on my communication as well, I never said I'd spend time to clean up the code, I only offered my autogenerated version as a possible starting point if someone else where to pick this up. I had mistakenly also thought Wallace had the opportunity to try to finish the translation so I left it at that.

If someone without Pro would take the time to create the translation, I'm happy to quickly doublecheck/polish the work and submit it to the repo. IMO the work is somewhat simple, just tedious and repetative, and not the best use of my time. It's also the reason I made the script, to automate part of the process, altough it only work ls on the better structured sheets.

**Comment by Lockbox313** [comment]
> "several months have passed without news"
Saying this means you expect and are entitled to something and you have not received it so therefore this is a complaint. 

@Anduh Running it  through again would be pointless since you are using the Roll20 github  code which has structure problems and your generator is still stripping > of various tags. I have again went through the HTML  and fixed some of it but as this stage all the skins after the Hollow are still trashed and once this is done work will probably have to been  done on the generated translation.json.  I have no more time to work on this today but here is an invite to the test game  https://app.roll20.net/join/4869171/AU04fQ

**Comment by JBeighel** [comment]
Shoot, I didn't know you both were working on this.  I had started down a different path of creating a translation.json file working for it.  

I don't know that its helpful since I was going a very different route, but the code I had is here:
https://github.com/JBeighel/roll20-character-sheets/tree/master/Monsterhearts-2

I don't have the Witch Hexes/Infernal Bargains, Experience, Advancement, or macros done yet.  The rest I'm pretty sure is covered, though the translation file likely could be organized better.

**Comment by Anduh** [comment]
> Shoot, I didn't know you both were working on this. I had started down a different path of creating a translation.json file working for it.

I think **Lockbox313** just did something quick just now since joining this discussion. I guess this will get done now that both you and **Lockbox313** have interest in getting the translation done. Communicate with each-other and good luck.

**Comment by JBeighel** [comment]
> **Edit:** IMO the `text-` prefix on the tag names is a bit redundant and just makes them unnecessarily verbose, but don't go and change it now that it's the way it's currently.

You're right, its an old habit to distinguish string length.  It likely doesn't help much here since there aren't many useful distinctions to make.

**Comment by Anduh** [comment]
> You're right, its an old habit to distinguish string length. It likely doesn't help much here since there aren't many useful distinctions to make.

Many sheets name the tag-name so it's understandable on their own, [Starfinder](https://github.com/JBeighel/roll20-character-sheets/blob/master/Starfinder%20by%20Roll20/translation.json) is one such example, but on the other hand I think your `move1desc`,`move2desc` etc have benefits from a sheet structure standpoint, so either way is great and internal consistency is always the smart thing to do.

**Comment by Lockbox313** [comment]
@JBeighel Are you working on Tremulus sheet or Monsterhearts 2 sheet?

**Comment by JBeighel** [comment]
Both, I just created the Tremulus sheet.  I borrowed a lot of HTML from the Monsterhearts so I'm familiar with it and thought I'd pitch in on this request for the Monsterhearts sheet.

I was re-using the tools for adding translation stuff to the Tremulus sheet to create similar tags in the Monsterhearts sheet.  Which is what I was showing in that previous repository link.

**Comment by Lockbox313** [comment]
I would go ahead with your version because  fixing the generated html and the generated  translation.josn would probably take longer than finishing off your version.

**Comment by JBeighel** [comment]
Ok, I can keep working on that.  It will stall a bit in the macros, they use external images to show some of the text (Roll, Success, and Failure I think).  

Do you know what was used to generate the HTML?  I don't do much HTML work, so I'm in the stone age with just a text editor.  Figured it might be good to look into modern tools.

**Comment by Lockbox313** [comment]
You things like HOT and the scrolling background?
 I use a text editor but then again that is how I started in the 80s'

**Comment by Lockbox313** [comment]
You do not need generate any html.  Look at the roll template at the bottom of the html file there area bunch of <img> tags.you justneed to change the src attribute to point your own image.

**Comment by JBeighel** [comment]
Making the image would be the trick.  I'm not good enough at image work to have it match the theme.  I guess I can create the tag and leave it open to creating an image or just inserting text.

**Comment by Lockbox313** [comment]
What images do you need?

**Comment by JBeighel** [comment]
I guess I'm thinking about it wrong, I don't need them it'd be the translator who'd have to create these in the other languages:
http://i.imgur.com/Y4a3NF3.png
http://i.imgur.com/ijSYmKZ.png
http://i.imgur.com/8IqLjkG.png
http://i.imgur.com/IJBloDk.png
http://i.imgur.com/Z4MvPRa.png

**Comment by Lockbox313** [comment]
I would not worry about the images for the Monsterhearts 2 translations. I don't think the  translation system can do anything about switching out images based on language anyway.  But you would probably want to change the images for  your Tremulus sheet.

**Comment by JBeighel** [comment]
Yup, that I did do.  Just have it showing the text.

**Comment by Lockbox313** [comment]
Ok Good luck pm at https://app.roll20.net/users/720651/ if you think I could help.

**Comment by JBeighel** [comment]
Finished the work on this one.  Submitted a pull request to bring in the changes.  Guess we'll see where it goes next.
